From: Jiri Slaby <jslaby@suse.cz>
Date: Fri, 15 Feb 2019 08:05:58 +0100
Subject: Revert "exec: load_script: don't blindly truncate shebang string"
Patch-mainline: not yet, under discussion
References: shebang regression

This reverts commit c3b081f9e2e3377af8c28336e23efab606268eb3, upstream
commit 8099b047ecc431518b9bb6bdbba3549bbecdc343. It changes the
behaviour in userspace with respect to shebang handling. Revert until
fixed.

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 fs/binfmt_script.c |   10 +++-------
 1 file changed, 3 insertions(+), 7 deletions(-)

--- a/fs/binfmt_script.c
+++ b/fs/binfmt_script.c
@@ -42,14 +42,10 @@ static int load_script(struct linux_binp
 	fput(bprm->file);
 	bprm->file = NULL;
 
-	for (cp = bprm->buf+2;; cp++) {
-		if (cp >= bprm->buf + BINPRM_BUF_SIZE)
-			return -ENOEXEC;
-		if (!*cp || (*cp == '\n'))
-			break;
-	}
+	bprm->buf[BINPRM_BUF_SIZE - 1] = '\0';
+	if ((cp = strchr(bprm->buf, '\n')) == NULL)
+		cp = bprm->buf+BINPRM_BUF_SIZE-1;
 	*cp = '\0';
-
 	while (cp > bprm->buf) {
 		cp--;
 		if ((*cp == ' ') || (*cp == '\t'))
